firedrake/linear_solver.py

```diff
@@ -133,14 +134,21 @@ def solve(self, x, b):
         if len(self._W) > 1 and self.near_nullspace is not None:
             self.near_nullspace._apply(self._W.dof_dset.field_ises, near=True)
         if self.A.has_bcs:
-            b_bc = self._b
-            # rhs = b - action(A, zero_function_with_bcs_applied)
-            b_bc.assign(b - self._Abcs)
-            # Now we need to apply the boundary conditions to the "RHS"
-            for bc in self.A.bcs:
-                bc.apply(b_bc)
-            # don't want to write into b itself, because that would confuse user
-            b = b_bc
+            if isinstance(self.A.a, slate.Tensor):
+                # Don't need action
+                b_bc = self._b
+                b_bc.assign(b)
+                for bc in self.A.bcs:
+                    bc.apply(b_bc)
```

Comment:
I think it's short-sighted to think you're only going to eliminate DG arguments: You could imagine a putative extension that does the static-condensation for CG (eliminating interior degrees of freedom), for example.